src/test/java/com/googlecode/download/maven/plugin/internal/WGetTest.java

```diff
@@ -48,29 +80,195 @@ public void setUp() throws Exception {
+        }
+    }
+
+    @Test
+    public void testCacheInNotADirectory() throws MojoExecutionException, IllegalAccessException {
+        try {
+            createMojo(m -> {
+                setVariableValueToObject(m, "overwrite", true);
+                File notADirectory = mock(File.class);
+                when(notADirectory.getAbsolutePath()).thenReturn("/nonExistingDirectory");
+                when(notADirectory.exists()).thenReturn(true);
+                when(notADirectory.isDirectory()).thenReturn(false);
+                setVariableValueToObject(m, "cacheDirectory", notADirectory);
+            }).execute();
+            fail();
+        } catch (MojoFailureException e) {
+            assertThat(e.getMessage(), containsString("cacheDirectory is not a directory"));
+        } finally {
+            cleanUp();
+        }
+    }
+
+    @Test
+    public void testCacheNotWrittenToIfFailed() throws MojoExecutionException, MojoFailureException, IllegalAccessException {
+        try {
+            createMojo(m -> setVariableValueToObject(m, "overwrite", true)).execute();
+            assertThat("Cache directory should remain empty if quit abruptly", cacheDirectory.list(),
+                    Matchers.emptyArray());
+        } finally {
+            cleanUp();
+        }
+    }
+
+    @Test
+    public void testReadingFromCache() throws Exception {
+        try {
+            // first, let's download something ;)
+            Mojo firstMojo = createMojo(mojo -> {
+                setVariableValueToObject(mojo, "overwrite", true);
+                setVariableValueToObject(mojo, "wagonManager", mock(WagonManager.class));
+
+                // stub maven session
+                class MavenSessionStub extends MavenSession {
+                    MavenSessionStub() {
+                        super(null, mock(MavenExecutionRequest.class), null, new LinkedList<>());
+                    }
+                }
+                setVariableValueToObject(mojo, "session", new MavenSessionStub());
+            });
+            HttpClientBuilder clientBuilder = mockClientBuilder("Hello, world!");
+            try (MockedStatic<HttpClientBuilder> httpClientBuilder = mockStatic(HttpClientBuilder.class)) {
+                httpClientBuilder.when(HttpClientBuilder::create).thenReturn(clientBuilder);
+
+                firstMojo.execute();
+            }
+
+            // now, let's try to read that from cache
+            createMojo(mojo -> setVariableValueToObject(mojo, "overwrite", true)).execute();
+
+            // now, let's read that file
+            Files.readAllLines(outputFile.toPath()).forEach(string -> assertThat(string, is("Hello, world!")));
+        } finally {
+            cleanUp();
+        }
+    }
+
+    @Test
+    public void testCacheRetainingValuesFromTwoConcurrentCalls() throws Exception {
+        File secondOutputFile = File.createTempFile("wget-test-", "");
+        try {
+            WGet longRunningMojo = createMojo(mojo -> {
```

Comment:
@ajarmoniuk not sure I understand why this mojo is long running. Could you, please, explain in javaoc?

Reply:
Sure, will do. However, the reason is simple: it finishes later because its input stream contains more bytes to process.